### PR TITLE
to_hf.py + PromptFormatter: produce vLLM-ready SpeechLM checkpoints with backbone-native chat_template

### PR DESCRIPTION
**Pull request by DongjiGao**
## Background

Reopens [#15617](https://github.com/NVIDIA-NeMo/NeMo/pull/15617) after its base branch (`speechlm2-with-nemo-automodel-merge`) was deleted; now rebased on `main`. Incorporates @pzelasko's review comments from the closed thread.

## Summary

Produces a single checkpoint artifact that serves both HF / NeMo SALM inference and vLLM (SpeechLM plugin) without an opt-in flag, and eliminates the drift between our hand-rolled jinja and the backbone LLM's canonical chat_template.

### Commit 1 - PromptFormatter uses backbone `chat_template` as source of truth

Replaces the hardcoded `TEMPLATE` dict + `to_jinja()` reconstruction in `QwenPromptFormatter` and `NemotronNanoV3PromptFormatter` with a shared base helper that delegates to `tokenizer.apply_chat_template` on the backbone's native jinja. Training and exported-ckpt inference now share one jinja string shipped by the backbone's HF repo - no more train/infer drift from reconstruction mismatches.

- `formatter.py`: add `AUDIO_TOKEN`, `_turn_to_openai_message`, `_is_reasoning_divergence`, `_encode_dialog_from_chat_template`. Drop the abstract `to_jinja` stub.
- `qwen.py`: `QwenPromptFormatter.encode_dialog` delegates; hardcoded `TEMPLATE`, `INFERENCE_PREFIX` and `to_jinja` removed. `Qwen3PromptFormatter` (reasoning, out of refactor scope) unchanged.
- `nemotron_nano_v3.py`: `encode_dialog` delegates and forwards `enable_thinking` to the backbone jinja. Default is `False`; reasoning training raises a clear 'not supported' error because the prefix-based loss mask can't align with the template's fresh `<think>` generation prompt.

### Commit 2 - `to_hf.py` unconditionally produces vLLM-ready checkpoints

- Removes the opt-in `vllm: bool = False` flag; every export is vLLM-ready (addresses @pzelasko's review comment on #15617).
- `_detect_vllm_architecture` now raises `ValueError` on missing `architectures` instead of silently defaulting (addresses the 'broad except' review comment).
- `_try_prepare_for_vllm` wrapper downgrades a `ValueError` to a warning so legacy SALM callers loading the same dir still get a valid HF-only checkpoint.
- Tokenizer saved from the backbone (brings its native `chat_template`) + augmented with `<|audio|>`. `extra_special_tokens` normalized to dict (HF loaders crash on the list form `transformers` writes for `additional_special_tokens`).
- For reasoning backbones (nemotron-nano-v3), the exported `chat_template`'s `enable_thinking` default is flipped to `False` so vLLM's request-time render matches training-time render; otherwise vLLM silently prepends `<think>` and WER regresses from 1.57 to 5.92.
- `hf_hub.py`: `setdefault('model_type', ...)` and `setdefault('architectures', [...])` in `HFHubMixin.save_pretrained` so NeMo-saved checkpoints carry the metadata vLLM / transformers need.

## Architecture mapping

| LLM backbone | Architecture name |
| ------------ | ----------------- |
| Hybrid (NemotronH) | `NeMoSpeechLMHybridForConditionalGeneration` |
| Standard (Qwen3, etc.) | `NeMoSpeechLMForConditionalGeneration` |

## Usage

```bash
torchrun --nproc-per-node=8 --nnodes=4 \
    examples/speechlm2/to_hf.py \
    class_path=nemo.collections.speechlm2.models.salm_automodel.SALMAutomodel \
    ckpt_path=/path/to/checkpoint.ckpt \
    ckpt_config=/path/to/exp_config.yaml \
    output_dir=/path/to/hf_output
```

No `vllm=True` flag any more - output is always vLLM-ready.

## Test plan

- [x] Parity test against backbone jinja for Qwen2.5-7B-Instruct and Nemotron-3-Nano-30B-A3B-BF16; rendered text byte-identical to the previous Python + to_jinja pipeline when `enable_thinking=False`.
- [x] Mask-construction invariants on training turns: `len(context) + len(answer) == len(input)` and `sum(mask) == len(answer)` across Qwen / Nemotron, plain / multi-turn, with / without system, multimodal list content.
- [x] End-to-end vLLM librispeech-pc eval on re-exported nemotron-nano-v3 ckpt: **WER 1.57 (== baseline)**. Without the `enable_thinking` default flip, WER regressed to 5.92 - regression confirmed, fix verified.
- [x] End-to-end NeMo SALM librispeech-pc eval on same ckpt: **WER 1.62, 0 empty hypotheses** (previous pipeline had 75 empties and 5.50 WER-all).
- [x] Backward compat: legacy SALM caller exports fine -- `_try_prepare_for_vllm` downgrades missing-prereq `ValueError` to a warning.
- [ ] CI: add 'Run CICD' label.

Made with Cursor

### CODE REVIEW

File: nemo/collections/common/prompts/formatter.py [hunk 1]
```diff
@@ -201,6 +201,120 @@ def __init_subclass__(cls, **kwargs) -> None:
...
+                messages,
+                tokenize=False,
+                add_generation_prompt=True,
+                **template_kwargs,
+            )
+            input_ids = torch.tensor(self._apply_tokenizer(text), dtype=torch.long)
+            return {"input_ids": input_ids, "context_ids": input_ids}
+
+        # Training: render twice (full dialog + context without last turn) so the
+        # tail of the full rendering becomes the loss-masked answer tokens.
```
Comment (DongjiGao) [inline nemo/collections/common/prompts/formatter.py:290]: not sure if there's a better way to do this

File: examples/speechlm2/to_hf.py [hunk 2]
```diff
@@ -158,14 +275,10 @@ def main(cfg: HfExportConfig):
 
     strategy_cfg = full_cfg.get("trainer", {}).get("strategy", {})
 
-    _is_torchrun = "RANK" in os.environ
-    if _is_torchrun and dist.is_available() and not dist.is_initialized():
+    if dist.is_available() and not dist.is_initialized():
```
Comment (pzelasko) [inline examples/speechlm2/to_hf.py:295]: Why are we removing `_is_torchrun`?

File: nemo/collections/common/prompts/formatter.py [hunk 3]
```diff
@@ -201,6 +201,120 @@ def __init_subclass__(cls, **kwargs) -> None:
                 ), f"{ERR} Slot '{slot}' not found in template '{template}' for role '{role}'"
         super().__init_subclass__(**kwargs)
 
+    AUDIO_TOKEN = "<|audio|>"
```
Comment (pzelasko) [inline nemo/collections/common/prompts/formatter.py:204]: Why is this defined globally? What happens when the model defines a different audio placeholder?

File: nemo/collections/common/prompts/formatter.py (hunk 1 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:290]: When we have:
```python
messages = [
  {"role": "user", "content": "..."},
  {"role": "assistant", "content": "..."},
  {"role": "user", "content": "..."},
  {"role": "assistant", "content": "..."},
]
```

the mask should be `[0, 0, 0, ..., 1, 1, 1, ..., 0, 0, 0, ..., 1, 1, 1]` to be compatible with current PromptFormatter implementation. This is not possible with `apply_chat_template` approach in the current form.

**Review by pzelasko — CHANGES_REQUESTED** [review]
I like your thinking about using chat template directly but for the reasons below I don't think it is feasible in the current shape. 

Let's approach this from two angles.

1. In this PR don't add chat template. Revert the changes from nemotron and qwen prompt formatter. Then fix `Nemotron...PromptFormatter` so that it's logic is equivalent to that of jinja, but using PromptFormatter. 
2. We can add `ChatTemplatePromptFormatter` in a separate PR which would be auto-selected when `prompt_format: "tokenizer_chat_template"` and make it generic for HF models.

**Comment by pzelasko** [comment]
This PR also needs unit test coverage of the changes.

**Comment by DongjiGao** [comment]
> I like your thinking about using chat template directly but for the reasons below I don't think it is feasible in the current shape.
> 
> Let's approach this from two angles.
> 
> 1. In this PR don't add chat template. Revert the changes from nemotron and qwen prompt formatter. Then fix `Nemotron...PromptFormatter` so that it's logic is equivalent to that of jinja, but using PromptFormatter.
> 2. We can add `ChatTemplatePromptFormatter` in a separate PR which would be auto-selected when `prompt_format: "tokenizer_chat_template"` and make it generic for HF models.

Could you please explain a bit why it is not feasible in current shape? Thanks!

File: examples/speechlm2/to_hf.py (hunk 2 above)
Comment (DongjiGao) [reply examples/speechlm2/to_hf.py]: Reverted.

File: nemo/collections/common/prompts/formatter.py (hunk 3 above)
Comment (DongjiGao) [reply nemo/collections/common/prompts/formatter.py]: Fix. Now directly read from training config. Raise error if not exist.

File: nemo/collections/common/prompts/qwen.py [hunk 4]
```diff
@@ -24,7 +24,10 @@
 class QwenPromptFormatter(PromptFormatter):
     NAME = "qwen"
     OUTPUT_ROLE = "assistant"
-    INFERENCE_PREFIX = f"{QWEN_BOT}assistant\n"
+    # Qwen3's chat_template injects this when enable_thinking=False; mirror it
+    # here so the SpeechLM fine-tune matches pre-training "reasoning-off" mode.
+    NO_THINK_PREFIX = "<think>\n\n</think>\n\n"
+    INFERENCE_PREFIX = f"{QWEN_BOT}assistant\n{NO_THINK_PREFIX}"
```
Comment (pzelasko) [inline nemo/collections/common/prompts/qwen.py]: I'm afraid this will be a breaking change for `canary-qwen-2.5b`. Let's roll back and add a separate `Qwen3PromptFormatter` in a follow up PR to reduce the scope here.

Comment (DongjiGao) [reply nemo/collections/common/prompts/qwen.py]: Yes you are right. I have verified that this would degrade canary-qwen-2.5 as it did not see the thinking tag during training. Reverting it.

**Review by pzelasko — APPROVED** [review]
LGTM

File: nemo/collections/common/prompts/formatter.py (hunk 1 above)
Comment (DongjiGao) [reply nemo/collections/common/prompts/formatter.py]: Will solve this in a separate PR.

**Review by pzelasko — APPROVED** [review]

File: examples/speechlm2/to_hf.py [hunk 5]
```diff
@@ -106,14 +107,134 @@ def save_hf_checkpoint(model: torch.nn.Module, state_dict: dict, cfg: HfExportCo
...
+    output_dir = Path(output_dir)
+    pretrained_llm = model_cfg.get("pretrained_llm", "")
+    if not pretrained_llm:
+        raise ValueError("model config has no 'pretrained_llm'; cannot load tokenizer for vLLM")
+
+    # ``model.audio_locator_tag`` is the SoT for the audio placeholder;
+    # fail loud rather than default, since a mismatch is silent at inference.
+    audio_token = model_cfg.get("audio_locator_tag")
+    if not audio_token:
+        raise ValueError("model config has no 'audio_locator_tag' (set it in the training YAML).")
```
Comment (stevehuang52) [inline examples/speechlm2/to_hf.py:165]: I think we shouldn't make `audio_locator_tag` as a required field for this script, since this script is used for any model in speechlm2 collection, and we shouldn't assume that this param will always be set in the config. If this script is `salm_to_hf.py`, it's fine to have such assumption, but this script is for more geneal speechlm models. @DongjiGao our streaming speechlm doesn't have this param in cfg.
